tests/Microsoft.Identity.Test.Unit/pop/PopAuthenticationSchemeTests.cs

```diff
@@ -129,35 +128,57 @@ public async Task ValidateKeyExpirationAsync()
                 testClock.TestTime = DateTime.UtcNow;
                 var provider = PoPProviderFactory.GetOrCreateProvider(testClock);
 
-                await app.AcquireTokenForClient(TestConstants.s_scope)
+                var result = await app.AcquireTokenForClient(TestConstants.s_scope)
                     .WithProofOfPossession(popConfig)
                     .ExecuteAsync(CancellationToken.None)
                     .ConfigureAwait(false);
+                var initialToken = result.AccessToken;
 
                 var JWK = provider.CannonicalPublicKeyJwk;
-                //Advance time 7 hours. Should still be the same key
+                //Advance time 7 hours. Should still be the same key and token
                 testClock.TestTime = testClock.TestTime + TimeSpan.FromSeconds(60 * 60 * 7);
 
                 provider = PoPProviderFactory.GetOrCreateProvider(testClock);
-                await app.AcquireTokenForClient(TestConstants.s_scope)
+                result = await app.AcquireTokenForClient(TestConstants.s_scope)
                     .WithProofOfPossession(popConfig)
                     .ExecuteAsync(CancellationToken.None)
                     .ConfigureAwait(false);
 
+                var handler = new JwtSecurityTokenHandler();
+                var jsonToken = handler.ReadJwtToken(result.AccessToken);
+
+                var jwtDecoded = Base64UrlHelpers.Decode(jsonToken.EncodedPayload);
+                var jObj = JObject.Parse(jsonToken.Payload.First().Value.ToString());
+                var at = jObj["jwk"]["n"].ToString();
+
+                Assert.AreEqual(GetAccessTokenFromPopToken(result.AccessToken), GetAccessTokenFromPopToken(initialToken));
                 Assert.IsTrue(JWK == provider.CannonicalPublicKeyJwk);
+                Assert.IsTrue(result.AuthenticationResultMetadata.TokenSource == TokenSource.Cache);
+
                 //Advance time 2 hours. Should be a different key
                 testClock.TestTime = testClock.TestTime + TimeSpan.FromSeconds(60 * 60 * 2);
 
                 provider = PoPProviderFactory.GetOrCreateProvider(testClock);
-                await app.AcquireTokenForClient(TestConstants.s_scope)
+                result = await app.AcquireTokenForClient(TestConstants.s_scope)
                     .WithProofOfPossession(popConfig)
                     .ExecuteAsync(CancellationToken.None)
                     .ConfigureAwait(false);
 
                 Assert.IsTrue(JWK != provider.CannonicalPublicKeyJwk);
+                Assert.IsTrue(result.AuthenticationResultMetadata.TokenSource == TokenSource.Cache);
```

Comment:
I was thinking about shat should happen here. After the key rotates and another call is made to the token cache, the same token is returned but the key is updated. i would imagine the key rotating should trigger a new call to the token endpoint right?

@bgavrilMS

Reply:
Is this the filter (key Id?)?
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/f9083e227400c75054ba594256ce1246092d8f4c/src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs#L652

Makes sense - token with expired key ID should be treated like a normal expired token.

Reply:
Yes, correct, if the key rotates, the at_cnf in the cache no longer matches it. So the `FindAccessToken` cache operation correctly results in a cache miss and MSAL makes another request to the /token endpoint.

Reply:
So the `TokenSource` should be TokenSource.IdentityProvider here. 
In fact MSAL does the right thing, but this test fails to correctly simulate the fact that time has passed.